### DISCUSSION TITLE
antsRegistrationSyn mean BOLD and steps

### DISCUSSION THREAD

**Discussion by ejurban** [ANTsX/ANTs#1344] [Q&A] [answered]
Hello,

I am trying to register my EPI images to standard space. I have a few questions that I haven't been able to find answers to online (I apologize if I have missed the documentation or instructions for best practices in my search). I would like to register each of my 4D bold time series to each subject's T1w structural, then to standard MNI space. I have a few specific questions.

1) I first tested antsRegistrationSyn.sh on one subject's 4D bold data (moving) to their T1w structural (fixed) and found it worked very well. I then read in the help that it is recommended that this step should be done with a 3D mean of the bold time series. I tried the same step with a 3D mean and found the registration worked similarly well in a similar amount of time. My question for you: in both cases, would the next step be to use ApplyTransform on the 4D timeseries? I'm guessing yes, since even the result of my 4D registration gave me a 3D image. If the next step is to ApplyTransform either way, and the step took a similar amount of time, why do you recommend running the registration using a 3D mean first instead of using the whole timeseries?

2) Ultimately, I would like to transform the bold data into MNI space. Should I next use antsRegistrationSyn to register the result of 1) above to the MNI template? And THEN run ApplyTransform? Or should I run ApplyTransform to the result of 1) and register that to the MNI template (followed by another ApplyTransform)? 

Thanks very much!
Emily

**Comment by ntustison** [comment]
First, start with ``antsRegistrationSyNQuick``.  It's much faster and probably sufficient.

Second, have you looked at [this example](https://stnava.github.io/fMRIANTs/)?  It will probably answer your questions.

**Reply by ejurban** [reply]
Thanks, @ntustison. I will give antsRegistraionSyNQuick a shot instead. 

I took a look at the example but see it is for using ANTsR. I'd like to do this using shell commands, as that fits into my lab's existing pipeline. I am still confused as to the order of operations here, particularly because I am new to ANTs and have only ever used FSL. Would you be able to describe the order of how registering a BOLD timeseries to MNI space would work using the different ANTs commands? Would it look something like below:

1. Find mean of BOLD series (in FSL, this would be: fslmaths bold.nii.gz -Tmean mean_bold.nii.gz)
2. use antsRegistrationSyNQuick.sh to map T1 (moving) to MNI template (fixed)
3. use antsRegistrationSyNQuick.sh to map mean_bold (moving) to the result of the second step (fixed)
4. use antsApplyTransform to apply mapping to from the third step to the entire bold time series

Thanks for your help,
Emily

**Comment by ntustison** [comment]
Take a look again---the example doesn't involve ANTsR.

But, yes, the steps that you outline are correct.

**Reply by ejurban** [reply]
Ah, right. Sorry about that- I clicked the link that sent me to the full pipeline example using ANTsR. 

Thanks for confirming!

**Reply by ejurban** [reply]
Thanks again for your help, @ntustison. I have spent quite a bit of time looking through the example you linked and emulating for my use case. I feel I have the general pipeline down but have two additional questions I am hoping you might be able to help me with (or may be able to point me to some additional resources). The questions are at the bottom of the post, but first, some background:

The example you sent has two transformations: motion correction, then registering bold data to a template. It also applies the data back to the 4D bold timeseries. What I'm trying to do is a little different in that 1) I'm not motion correcting (I'll be doing that at another stage); 2) I'm registering the bold data to structural data AND to an MNI template; and 3) I am not applying to 4D time series just yet. Given these differences I'm just a bit unsure whether my updates to the script you sent are correct. (PS- I decided to use SyN and not SyNQuick as it does much better with my data)

Here are my steps:
1) find average of bold data (I call "example_func")
2) Use antsRegistrationSyN to register example_func to the highres image (subject's T1 structural)
3) Use antsRegistrationSyN to register highres to standard MNI template
4) Use antsApplyTransforms to combine the affine (highres2standard_0GenericAffine.mat) and non-linear (highres2standard_1Warp.nii.gz) transformations from step 3 into one transformation (highres2standard_CollapsedWarp.nii.gz)
5) Use antsApplyTransforms to apply the transformations from step 2 and step 4 to the example_func image.

It is step 5 I am unsure about. Here is the command I am using:
antsApplyTransforms -d 3 -o $example_func2standard_1WarpCollapsedWarp.nii.gz \
  -t $highres2standard_CollapsedWarp.nii.gz -t $example_func2highres_1Warp.nii.gz \
  -r $standard -i $example_func

My two main questions:
1) Is it possible/advisable to have an additional step before step 5 that first incorporates all the transformations from Step 2 and Step 4, before applying to example_func? I have tried this (basically just step 5 without the -i $example_func) and I get an error saying "Segmentation fault (core dumped)". I like the idea of this step, as it would allow me to save the overall transformation to be applied to the data at a later stage, rather than just applying to the data and saving the end result. 
2) If I just jump straight to transforming example_func using the transformations from Step 2 and Step 4 (my current pipeline outlined above), for the transformation from step 2 is it correct to use the $example_func2highres_1Warp.nii.gz? Or should I be using $example_func2highres_0GenericAffine.mat instead or in addition to? The result LOOKS good, but I guess I'm not sure why I'm using 1Warp instead of 0GenericAffine (perhaps 0GenericAffine is incorporated into 1Warp?)

Thank you very much for any help you're able to provide!
Emily

**Comment by ntustison** [comment, marked as answer]
> Is it possible/advisable to have an additional step before step 5 that first incorporates all the transformations from Step 2 and Step 4, before applying to example_func? I have tried this (basically just step 5 without the -i $example_func) and I get an error saying "Segmentation fault (core dumped)". I like the idea of this step, as it would allow me to save the overall transformation to be applied to the data at a later stage, rather than just applying to the data and saving the end result.

To compose transforms, you must specify ``-o [$example_func2standard_1WarpCollapsedWarp.nii.gz,1]``.

> If I just jump straight to transforming example_func using the transformations from Step 2 and Step 4 (my current pipeline outlined above), for the transformation from step 2 is it correct to use the $example_func2highres_1Warp.nii.gz? Or should I be using $example_func2highres_0GenericAffine.mat instead or in addition to? The result LOOKS good, but I guess I'm not sure why I'm using 1Warp instead of 0GenericAffine (perhaps 0GenericAffine is incorporated into 1Warp?)

By default, the output of the antsRegistration programs maintains separation between the linear and deformable parts of the total output transform.  Unless you do something special, this output will typically be ``${outputPrefix}0GenericAffine.mat`` and ``${outputPrefix}1Warp.nii.gz``/``${outputPrefix}1InverseWarp.nii.gz``.  To apply the set of transforms to warp the moving image to the fixed space, you would use ``antsApplyTransforms`` as follows:

```
$ antsApplyTransforms ... -i ${fixed} -r ${moving} -t ${outputPrefix}1Warp.nii.gz -t ${outputPrefix}0GenericAffine.mat ...
```

Note that the order matters, specifically transforms are read in reverse order as specified in the command.  Therefore, to do the inverse of the above, you would specify:

```
$ antsApplyTransforms ... -i ${moving} -r ${fixed} -t [${outputPrefix}0GenericAffine.mat,1]  -t ${outputPrefix}1InverseWarp.nii.gz ...
```

**Reply by ejurban** [reply]
This is extremely helpful, thanks very much @ntustison! This worked beautifully and makes much more sense now. All the best,
Emily
